dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/programrule/implementers/enrollment/AssignValueExecutor.java

```diff
@@ -0,0 +1,163 @@
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.programrule.implementers.enrollment;
+
+import java.util.List;
+import java.util.Optional;
+
+import lombok.RequiredArgsConstructor;
+
+import org.apache.commons.lang3.math.NumberUtils;
+import org.hisp.dhis.common.ValueType;
+import org.hisp.dhis.setting.SettingKey;
+import org.hisp.dhis.setting.SystemSettingManager;
+import org.hisp.dhis.system.util.MathUtils;
+import org.hisp.dhis.trackedentity.TrackedEntityAttribute;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Attribute;
+import org.hisp.dhis.tracker.domain.Enrollment;
+import org.hisp.dhis.tracker.domain.MetadataIdentifier;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.programrule.IssueType;
+import org.hisp.dhis.tracker.programrule.ProgramRuleIssue;
+import org.hisp.dhis.tracker.validation.ValidationCode;
+
+import com.google.common.collect.Lists;
+
+/**
+ * This executor assigns a value to a field if it is empty, otherwise returns an
+ * error
+ *
+ * @Author Enrico Colasante
+ */
+@RequiredArgsConstructor
+public class AssignValueExecutor implements RuleActionExecutor
+{
+    private final SystemSettingManager systemSettingManager;
+
+    private final String ruleUid;
+
+    private final String value;
+
+    private final String attributeUid;
+
+    private final List<Attribute> attributes;
+
+    @Override
+    public Optional<ProgramRuleIssue> executeEnrollmentRuleAction( TrackerBundle bundle, Enrollment enrollment )
+    {
+        Boolean canOverwrite = systemSettingManager
+            .getBooleanSetting( SettingKey.RULE_ENGINE_ASSIGN_OVERWRITE );
+        TrackedEntityAttribute attribute = bundle.getPreheat().getTrackedEntityAttribute( attributeUid );
+
+        Optional<Attribute> payloadAttribute = attributes.stream()
+            .filter( at -> at.getAttribute().isEqualTo( attribute ) )
+            .findAny();
+
+        if ( payloadAttribute.isEmpty() ||
+            Boolean.TRUE.equals( canOverwrite ) ||
+            isEqual( value, payloadAttribute.get().getValue(), attribute.getValueType() ) )
+        {
+            addOrOverwriteAttribute( enrollment, bundle );
+            return Optional.of( new ProgramRuleIssue( ruleUid, ValidationCode.E1310,
```

Comment:
We could create helper factories in `ProgramRuleIssue` like `warning()` and `error()`. This would at least remove the need for the `IssueType.WARNING` parameter.

It could also open up the use of varargs for the list arg.

So turn these instantiations into

```java
warning(ruleUid, E1310, attributeUid, value);
```